src/lightning/pytorch/callbacks/batch_size_finder.py

```diff
@@ -128,24 +128,25 @@ def __init__(
     def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: Optional[str] = None) -> None:
         if trainer._accelerator_connector.is_distributed:
             raise MisconfigurationException("The Batch size finder is not supported with distributed strategies.")
-
-        running_stage = trainer.state.stage
-        assert running_stage is not None
-        dl_source = getattr(trainer._data_connector, f"_{running_stage.dataloader_prefix}_dataloader_source")
-
         # TODO: check if this can be enabled (#4040)
-        if not trainer._data_connector._train_dataloader_source.is_module():
+        if not trainer.fit_loop._data_source.is_module():
             raise MisconfigurationException(
                 "The Batch size finder cannot be used with dataloaders passed directly to `.fit()`. Please disable"
                 " the feature or incorporate the dataloader into your LightningModule or LightningDataModule."
             )
 
         # TODO: Add support for multiple eval dataloader
         if stage != "fit":
-            dataloaders = dl_source.dataloader()
-            if isinstance(dataloaders, list) and len(dataloaders) > 1:
+            loop = trainer._active_loop
+            assert loop is not None
+            loop.setup_data()
+            combined_loader = loop._combined_loader
+            assert combined_loader is not None
```

Comment:
same about mypy

Reply:
mypy doesnt know that `setup_data` will set `loop._combined_loader`. We know this because we control the implementation, but this could be violated either in the future, or by someone hacking around the codebase.

In my opinion an `assert` is preferred, because we want to be defensive about this and cover the error.

I see `cast()` more useful to circumvent a limitation of the mypy engine. For example, because some feature hasn't been implemented yet.

We also use `assert` all around the codebase already.

Reply:
Currently we use a mixture of both. I'm fine with preferring assert in these cases though.